### PR TITLE
feat(DTFS2-6624): portal UI add read only limitation on UI elements

### PR DESCRIPTION
**Pull request by AlexBramhill**
# Introduction
This PR adds the read-only role to portal ui components, and associated tests.

# Resolution
This PR refactors several components to align with existing patterns. It adds more extensive component tests for role based access. It follows existing convention of checking the role in the component, as apposed to the js.

It focuses on the following:
- Clone a deal
- Abandon a deal
- Submit the deal further down the deal journey.
- Unable to change any attribute of the deal.
- Unable to issue a facility
- Unable to generate any reports (by removing the ability to navigate to the reports page -- routing 401 is handled by #1980).

# Misc
I've done some general tidying up including:
- Updating some test data to comply with cspell
- Updated roles to use a constant (note -- the routing validation has not been updated to use this, however #1980 will cover this change.

### CODE REVIEW

File: portal/component-tests/contract/components/abandon-deal-link.component-test.js [hunk 1]
```diff
@@ -4,9 +4,17 @@ const component = 'contract/components/abandon-deal-link.njk';
 const render = componentRenderer(component);
 
 describe(component, () => {
-  describe('when viewed by a maker', () => {
+  const makerRole = ['maker'];
+  const nonMakerRoles = ['checker', 'admin', 'read-only'];
+
+  function makerRoleTests(roleToCombineWithMaker) {
```
Comment (AlexBramhill) [inline portal/component-tests/contract/components/abandon-deal-link.component-test.js:10]: We probably don't need the roleToCombineWithMaker anymore

File: portal/component-tests/contract/components/requested-start-date-link.component-test.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+        wrapper
+          .expectText(`[data-cy="${facilityName}-change-or-confirm-cover-start-date-${facility._id}"]`)
+          .toRead('Start date confirmed');
+      });
+    });
+
+    describe('when hasNotConfirmedCoverStartDate is false', () => {
+      const hasConfirmedCoverStartDate = false;
+      it('should render no link and "Start date not confirmed"', () => {
```
Comment (lmarrai-sw) [inline portal/component-tests/contract/components/requested-start-date-link.component-test.js:74]:
```suggestion
      const hasConfirmedCoverStartDate = false;
      
      it('should render no link and "Start date not confirmed"', () => {
```

File: portal/component-tests/contract/components/requested-start-date-link.component-test.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+          .toLinkTo(`/contract/${deal._id}/${facilityName}/${facility._id}/confirm-requested-cover-start-date`, 'Confirm start date');
+      });
+    });
+  });
+
+  describe.each(nonMakerRoles)('when viewed with the role %s', (nonMakerRole) => {
+    const user = { roles: [nonMakerRole] };
+    describe('when hasConfirmedCoverStartDate is true', () => {
+      const hasConfirmedCoverStartDate = true;
+      it('should render no link and "Start date confirmed"', () => {
```
Comment (lmarrai-sw) [inline portal/component-tests/contract/components/requested-start-date-link.component-test.js:57]:
```suggestion
      const hasConfirmedCoverStartDate = true;
      
      it('should render no link and "Start date confirmed"', () => {
```

File: portal/component-tests/contract/components/requested-start-date-link.component-test.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        wrapper
+          .expectLink(`[data-cy="${facilityName}-change-or-confirm-cover-start-date-${facility._id}"]`)
+          .toLinkTo(`/contract/${deal._id}/${facilityName}/${facility._id}/confirm-requested-cover-start-date`, 'Confirm start date');
+      });
+    });
+  });
+
+  describe.each(nonMakerRoles)('when viewed with the role %s', (nonMakerRole) => {
+    const user = { roles: [nonMakerRole] };
+    describe('when hasConfirmedCoverStartDate is true', () => {
```
Comment (lmarrai-sw) [inline portal/component-tests/contract/components/requested-start-date-link.component-test.js:55]:
```suggestion
    const user = { roles: [nonMakerRole] };
    
    describe('when hasConfirmedCoverStartDate is true', () => {
```

File: portal/component-tests/contract/components/requested-start-date-link.component-test.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  const facility = { _id: '5f3ab3f705e6630007dcfb22' };
+  const facilityName = 'Loan';
+
+  const deal = { _id: '5f3ab3f705e6630007dcfb20' };
+
+  describe('when viewed as a maker', () => {
+    const user = { roles: [MAKER] };
+    describe('when hasConfirmedCoverStartDate is true', () => {
+      const hasConfirmedCoverStartDate = true;
+      it('should render a link and "Start date confirmed"', () => {
```
Comment (lmarrai-sw) [inline portal/component-tests/contract/components/requested-start-date-link.component-test.js:20]:
```suggestion
    const user = { roles: [MAKER] };
    
    describe('when hasConfirmedCoverStartDate is true', () => {
      const hasConfirmedCoverStartDate = true;
      
      it('should render a link and "Start date confirmed"', () => {
```

File: portal/component-tests/contract/components/loan-link.component-test.js [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+const component = 'contract/components/loan-link.njk';
+const render = componentRenderer(component);
+
+describe(component, () => {
+  const nonMakerRoles = [CHECKER, ADMIN, READ_ONLY];
+  const deal = { _id: '61f6fbaea2460c018a4189d7' };
+  const loanWithName = { _id: '61f6fbaea2460c018a4189da', name: 'Test Loan Name' };
+  const loanWithoutName = { _id: '61f6fbaea2460c018a4189db' };
+  function itShouldRenderLoanNameAndHyperlink(params, loan) {
```
Comment (lmarrai-sw) [inline portal/component-tests/contract/components/loan-link.component-test.js:13]:
```suggestion
  const loanWithoutName = { _id: '61f6fbaea2460c018a4189db' };
  
  function itShouldRenderLoanNameAndHyperlink(params, loan) {
```

File: portal/component-tests/contract/components/loan-link.component-test.js [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    describe('when loan.name is present', () => {
+      describe('when loan is editable', () => {
+        const params = {
+          user,
+          deal,
+          editable: true,
+        };
+        itShouldRenderLoanNameAndHyperlink(params, loanWithName);
+      });
+      describe('when loan is not editable', () => {
```
Comment (lmarrai-sw) [inline portal/component-tests/contract/components/loan-link.component-test.js:73]:
```suggestion
      });
      
      describe('when loan is not editable', () => {
```

File: portal/component-tests/contract/components/loan-link.component-test.js [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    describe('when loan.name is not present', () => {
+      describe('when loan is editable', () => {
+        const params = {
+          user,
+          deal,
+          editable: true,
+        };
+        itShouldRenderLoanRefNotEnteredAndHyperlink(params, loanWithoutName);
+      });
+      describe('when loan is not editable', () => {
```
Comment (lmarrai-sw) [inline portal/component-tests/contract/components/loan-link.component-test.js:92]:
```suggestion
      });
      
      describe('when loan is not editable', () => {
```

File: portal/component-tests/contract/components/loan-link.component-test.js [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    describe('when loan.name is present', () => {
+      describe('when loan is editable', () => {
+        const params = {
+          user,
+          deal,
+          editable: true,
+        };
+        itShouldRenderLoanName(params, loanWithName);
+      });
+      describe('when loan is not editable', () => {
```
Comment (lmarrai-sw) [inline portal/component-tests/contract/components/loan-link.component-test.js:114]:
```suggestion
      });
      
      describe('when loan is not editable', () => {
```

File: portal/component-tests/contract/components/loan-link.component-test.js [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+    describe('when loan.name is not present', () => {
+      describe('when loan is editable', () => {
+        const params = {
+          user,
+          deal,
+          editable: true,
+        };
+        itShouldRenderLoanRefNotEntered(params, loanWithoutName);
+      });
+      describe('when loan is not editable', () => {
```
Comment (lmarrai-sw) [inline portal/component-tests/contract/components/loan-link.component-test.js:133]:
```suggestion
      });
      
      describe('when loan is not editable', () => {
```

File: portal/component-tests/contract/components/contract-actions.component-test.js [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+
+      it('should be disabled for deals in any state', () => {
+        for (const deal of [...readyForCheckerApprovalDeals, ...draftAndFurtherInputRequiredDeals, ...otherDeals]) {
+          const wrapper = render({ user, dealFormsCompleted, deal });
+          wrapper.expectText('[data-cy="canProceed"]').notToExist();
+        }
+      });
+    });
+  }
+  function otherRoleTests(otherRole) {
```
Comment (lmarrai-sw) [inline portal/component-tests/contract/components/contract-actions.component-test.js:100]:
```suggestion
  }
  
  function otherRoleTests(otherRole) {
```

File: portal/component-tests/contract/components/contract-actions.component-test.js [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+
+      it('should be disabled for deals in all other states', () => {
+        for (const deal of [...readyForCheckerApprovalDeals, ...otherDeals]) {
+          const wrapper = render({ user, dealFormsCompleted, deal });
+          wrapper.expectText('[data-cy="canProceed"]').notToExist();
+        }
+      });
+    });
+  }
+  function checkerRoleTests() {
```
Comment (lmarrai-sw) [inline portal/component-tests/contract/components/contract-actions.component-test.js:68]:
```suggestion
  }
  
  function checkerRoleTests() {
```

**Review by lmarrai-sw — COMMENTED** [review]
Thank you for the great work @AlexBramhill, you've added really great tests here. 
Most of the comments are me suggesting line breaks so don't be put off by the number of comments!

File: e2e-tests/portal/cypress/e2e/journeys/read-only/fixtures/dealFullyCompleted.js [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+      value: '12345678',
+      currencySameAsSupplyContractCurrency: 'true',
+      interestMarginFee: '12',
+      coveredPercentage: '20',
+      minimumQuarterlyFee: '20',
+      premiumFrequency: 'Monthly',
+      premiumType: 'In advance',
+      dayCountBasis: '365',
+      currency: {
+        text: 'GBP - UK Sterling',
```
Comment (abhi-markan) [inline e2e-tests/portal/cypress/e2e/journeys/read-only/fixtures/dealFullyCompleted.js:114]: I believe there are constants of these values and others, can you refer to the constants for various values defined here please.

File: e2e-tests/portal/cypress/e2e/journeys/read-only/fixtures/dealWithInvalidBondCoverStartDate.js [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+  const invalidCoverStartDate = `${dateConstants.yesterdayUnix}000`;
+
+  const dealWithBadCoverStartDate = { ...dealReadyToSubmitForReview() };
+
+  dealWithBadCoverStartDate.bondTransactions.items[0].requestedCoverStartDate = invalidCoverStartDate;
+
+  dealWithBadCoverStartDate.bondTransactions.items[0]['coverEndDate-day'] = (dateConstants.oneMonthDay).toString();
+  dealWithBadCoverStartDate.bondTransactions.items[0]['coverEndDate-month'] = (dateConstants.oneMonthMonth).toString();
+  dealWithBadCoverStartDate.bondTransactions.items[0]['coverEndDate-year'] = (dateConstants.oneMonthYear).toString();
+  dealWithBadCoverStartDate.bondTransactions.items[0].facilityStage = 'Issued';
```
Comment (abhi-markan) [inline e2e-tests/portal/cypress/e2e/journeys/read-only/fixtures/dealWithInvalidBondCoverStartDate.js:14]: Constant please.

File: e2e-tests/portal/cypress/e2e/journeys/read-only/fixtures/dealWithInvalidBondCoverStartDate.js [hunk 15]
```diff
@@ -0,0 +1,17 @@
+const dealReadyToSubmitForReview = require('./dealReadyToSubmit');
+const dateConstants = require('../../../../../../e2e-fixtures/dateConstants');
+
+module.exports = () => {
+  const invalidCoverStartDate = `${dateConstants.yesterdayUnix}000`;
```
Comment (abhi-markan) [inline e2e-tests/portal/cypress/e2e/journeys/read-only/fixtures/dealWithInvalidBondCoverStartDate.js:5]: I know there are occurrences of this coding style in the source code, could you please add a TODO to standardise this (should be covered in a separate PR).

File: e2e-tests/portal/cypress/e2e/journeys/read-only/fixtures/submittedDealWithBondCoverStartDateInThePast.js [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+  const deal = { ...dealReadyToSubmitForReview() };
+
+  deal.details.submissionDate = dealSubmissionDate;
+
+  deal.bondTransactions.items[0].requestedCoverStartDate = coverStartDateBeforeDealSubmissionDate;
+
+  deal.bondTransactions.items[0]['coverEndDate-day'] = (dateConstants.oneMonthDay).toString();
+  deal.bondTransactions.items[0]['coverEndDate-month'] = (dateConstants.oneMonthMonth).toString();
+  deal.bondTransactions.items[0]['coverEndDate-year'] = (dateConstants.oneMonthYear).toString();
+  deal.bondTransactions.items[0].facilityStage = 'Issued';
```
Comment (abhi-markan) [inline e2e-tests/portal/cypress/e2e/journeys/read-only/fixtures/submittedDealWithBondCoverStartDateInThePast.js:17]: Constant, please go through all the fixtures introduced in this PR.

File: e2e-tests/portal/cypress/e2e/journeys/read-only/fixtures/template.json [hunk 17]
```diff
@@ -0,0 +1,168 @@
+{
+  "submissionType": "Automatic Inclusion Notice",
```
Comment (abhi-markan) [inline e2e-tests/portal/cypress/e2e/journeys/read-only/fixtures/template.json:2]: Please use constants

File: e2e-tests/portal/cypress/fixtures/users.js [hunk 18]
```diff
@@ -1,33 +1,26 @@
 const MOCK_USERS = require('../../../../utils/mock-data-loader/portal/users');
 const MOCK_TFM_USERS = require('../../../../utils/mock-data-loader/tfm/mocks/users');
 
-const BANK1_MAKER1 = MOCK_USERS.find((user) =>
-  (user.roles.includes('maker') && user.username === 'BANK1_MAKER1'));
+const BANK1_MAKER1 = MOCK_USERS.find((user) => user.roles.includes('maker') && user.username === 'BANK1_MAKER1');
```
Comment (abhi-markan) [inline e2e-tests/portal/cypress/fixtures/users.js:4]: Please use constants for `role` and `username`.

File: portal/component-tests/contract/components/abandon-deal-link.component-test.js (hunk 1 above)
Comment (abhi-markan) [reply portal/component-tests/contract/components/abandon-deal-link.component-test.js:10]: We could land up having a `maker` with combined role of a `checker` or and `admin`.

File: portal/component-tests/contract/components/bond-transactions-table.component-test.js [hunk 19]
```diff
@@ -41,100 +41,179 @@ describe(component, () => {
...
-      wrapper.expectText('[data-cy="bonds-table-header-ukef-facility-id"]').toRead('UKEF facility ID');
-      wrapper.expectText('[data-cy="bonds-table-header-status"]').toRead('Status');
-      wrapper.expectText('[data-cy="bonds-table-header-value"]').toRead('Value');
-      wrapper.expectText('[data-cy="bonds-table-header-stage"]').toRead('Stage');
-      wrapper.expectText('[data-cy="bonds-table-header-start-date"]').toRead('Start date');
-      wrapper.expectText('[data-cy="bonds-table-header-end-date"]').toRead('End date');
-      wrapper.expectText('[data-cy="bonds-table-header-action"]').toRead('Action');
+  const dealWithBondsThatCanChangeCoverDate = JSON.parse(JSON.stringify(deal));
+  dealWithBondsThatCanChangeCoverDate.status = 'Acknowledged';
+  dealWithBondsThatCanChangeCoverDate.bondTransactions.items[0].facilityStage = 'Issued';
```
Comment (abhi-markan) [inline portal/component-tests/contract/components/bond-transactions-table.component-test.js:46]: Constants please.

File: portal/component-tests/contract/components/clone-deal-link.component-test.js [hunk 20]
```diff
@@ -2,51 +2,58 @@ const componentRenderer = require('../../componentRenderer');
...
-        { _id: '61f6fbaea2460c018a4189d5', status: 'Abandoned' },
-        { _id: '61f6fbaea2460c018a4189d6', status: 'Acknowledged' },
-        { _id: '61f6fbaea2460c018a4189d7', status: 'Accepted by UKEF (without conditions)' },
-        { _id: '61f6fbaea2460c018a4189d8', status: 'Accepted by UKEF (with conditions)' },
-        { _id: '61f6fbaea2460c018a4189d9', status: "Ready for Checker's approval" },
-      ];
+  const nonMakerRoles = [CHECKER, ADMIN, READ_ONLY];
+
+  const deals = [
+    { _id: '61f6fbaea2460c018a4189d1', status: 'Draft' },
```
Comment (abhi-markan) [inline portal/component-tests/contract/components/clone-deal-link.component-test.js:11]: Can you look into using constants please?

File: portal/component-tests/partials/primary-navigation.component-test.js [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+    it('renders a users link', () => {
+      wrapper.expectLink('[data-cy="users"]').toLinkTo('/admin/users', 'Users');
+    });
+  }
+
+  function itDoesNotRenderAUsersLink() {
+    it('does not render a users link', () => {
+      wrapper.expectLink('[data-cy="users"]').notToExist();
+    });
+  }
```
[the remark this answers is not in the document]
Comment (abhi-markan) [reply portal/component-tests/partials/primary-navigation.component-test.js:42]: Happy for the common portal header links to be converted to a single Cypress command.

File: portal/server/routes/middleware/isMaker/index.js [hunk 22]
```diff
@@ -1,7 +1,9 @@
+const { MAKER } = require('../../../constants/roles');
+
 const isMaker = async (req, res, next) => {
```
Comment (abhi-markan) [inline portal/server/routes/middleware/isMaker/index.js:3]: Can we use one of the existing function `validateRole`?

File: portal/templates/_partials/primary-navigation.njk [hunk 23]
```diff
@@ -56,4 +60,20 @@
...
+  {
+    text: "Home",
+    href: "/dashboard",
+    active: primaryNav == "home",
+    attributes: {
+      "data-cy": "dashboard"
+    }
+  }
+  ]
+}) }}
```
[the remark this answers is not in the document]
Comment (abhi-markan) [reply portal/templates/_partials/primary-navigation.njk:77]: Yes that is correct.

File: portal/templates/_partials/primary-navigation.njk [hunk 24]
```diff
@@ -32,7 +36,7 @@
   ]
 }) }} 
 
-{% elif user %}
+{% elif isChecker or isMaker %}
```
Comment (abhi-markan) [inline portal/templates/_partials/primary-navigation.njk:39]: Does the UI allow a maker to be an `admin` as well so a combined `maker-admin` role is possible in the latest design?

File: portal/templates/contract/components/issue-or-delete-facility-link.njk [hunk 25]
```diff
@@ -1,37 +1,33 @@
...
-
-    {% elseif (userIsMaker and dealIsEditable) and (dealStatus === "Draft" or dealStatus === "Further Maker's input required") and not dealSubmissionDate %}
-      {% if facility.name %}
-        <a href="/contract/{{ dealId }}/{{ facilityName }}/{{ facility._id}}/delete" class="govuk-link" data-cy="{{ facilityName }}-delete-{{ facility._id }}" aria-label="Delete {{ facility.type }} {{ facility.name }}">Delete {{ facility.name }}</a>
-      {% else %}
-        <a href="/contract/{{ dealId }}/{{ facilityName }}/{{ facility._id}}/delete" class="govuk-link" data-cy="{{ facilityName }}-delete-{{ facility._id }}" aria-label="Delete {{ facility.type }} {{ facilityTableIndex }}">Delete {{ facility.type | lower | safe }}</a>
-      {% endif %}
+    {% endif %}
+    {# 3 #}
+  {% elseif (userIsMaker and dealIsEditable) and (dealStatus === "Draft" or dealStatus === "Further Maker's input required") and not dealSubmissionDate %}
```
Comment (abhi-markan) [inline portal/templates/contract/components/issue-or-delete-facility-link.njk:26]: Cognitive complexity is over 3, can this please be feed from the controller, any thoughts?

File: portal/templates/contract/components/issue-or-delete-facility-link.njk [hunk 26]
```diff
@@ -1,37 +1,33 @@
...
-      Issue facility
-    {% endif%}
+        Issue facility
+      {% endif %}
     </a>
-
-    {% elseif canOnlyViewIssueFacilityDetails and facility.issueFacilityDetailsProvided %}
+    {# 2 #}
+  {% elseif canOnlyViewIssueFacilityDetails or facility.canIssueOrEditIssueFacility %}
+    {% if facility.issueFacilityDetailsProvided %}
```
[the remark this answers is not in the document]
Comment (abhi-markan) [reply portal/templates/contract/components/issue-or-delete-facility-link.njk:22]: The naming convention is quite confusing.

File: portal/templates/contract/components/loan-link.njk [hunk 27]
```diff
@@ -1,27 +1,33 @@
...
+
   {% if loan.name %}
-    {% if params.editable %}
-      <a href="/contract/{{ dealId }}/loan/{{ loan._id}}/guarantee-details"
-         class="govuk-link"
-         data-cy="loan-bank-reference-number-link-{{loan._id}}"
-         title="Loan {{loan.name}}"
-         aria-label="Loan {{loan.name}}">
-         {{ loan.name | safe }}
+    {% if params.editable and params.user.roles.includes('maker') %}
```
Comment (abhi-markan) [inline portal/templates/contract/components/loan-link.njk:7]: Same as Luke's comment below.

File: portal/templates/contract/components/loan-transactions-table.njk [hunk 28]
```diff
@@ -29,9 +29,10 @@
           <tbody class="govuk-table__body">
 
             {% for loan in loans %}
+            {% set loanLinkParams = { params: params, loan: loan } %}
```
Comment (abhi-markan) [inline portal/templates/contract/components/loan-transactions-table.njk:32]: Should they be different variable names?

File: portal/templates/contract/components/loan-transactions-table.njk [hunk 29]
```diff
@@ -67,16 +68,15 @@
 
                   {% set componentData = {user: params.user, deal: params.deal, editable: params.editable, facility: loan, facilityName: 'loan', hasConfirmedCoverStartDate: hasConfirmedCoverStartDate, facilityTableIndex: loop.index} %}
 
-                  {% set userIsMaker = params.user.roles.includes('maker')%}
-
                   {% set dealSubmissionType = params.deal.submissionType %}
 
+                  {% set userIsMaker = params.user.roles.includes('maker')%}
+
                   {% set requestedCoverDateCanEdit = userIsMaker and loan.status !== "Maker's input required" and dealStatusAllowsRequestedCoverDateChange and dealSubmissionType === 'Manual Inclusion Application' and loan.facilityStage === 'Unconditional' and loan.issueFacilityDetailsSubmitted %}
```
Comment (abhi-markan) [inline portal/templates/contract/components/loan-transactions-table.njk:75]: I know this has not been included in your PR but can this also please be feed from the controller?

**Review by abhi-markan — CHANGES_REQUESTED** [review]
Great work, few comments to address please.

File: e2e-tests/portal/cypress/fixtures/users.js (hunk 18 above)
Comment (AlexBramhill) [reply e2e-tests/portal/cypress/fixtures/users.js]: updated `role`, i've left `username` as we're not currently doing this anywhere else and it would be good to do it all in one go to avoid fragmentation.

File: portal/templates/contract/components/issue-or-delete-facility-link.njk [hunk 30]
```diff
@@ -1,37 +1,33 @@
...
+        Facility issued
       {% else %}
-      Issue facility
-    {% endif%}
+        Issue facility
+      {% endif %}
     </a>
-
-    {% elseif canOnlyViewIssueFacilityDetails and facility.issueFacilityDetailsProvided %}
+    {# 2 #}
```
[the remark this answers is not in the document]
Comment (AlexBramhill) [reply portal/templates/contract/components/issue-or-delete-facility-link.njk]: removed-  was just helping with some debugging

File: portal/templates/contract/components/loan-link.njk [hunk 31]
```diff
@@ -1,27 +1,33 @@
...
     {% else %}
-      <span data-cy="loan-bank-reference-number-{{loan._id}}">
+      <span data-cy="loan-bank-reference-number-{{ loan._id }}">
         {{ loan.name | safe }}
       </span>
     {% endif %}
   {% else %}
-    {% if params.editable %}
-      <a href="/contract/{{ dealId }}/loan/{{ loan._id}}/guarantee-details" class="govuk-link" data-cy="loan-bank-reference-number-link-{{loan._id}}">Loan’s reference number not entered</a>
+    {% if params.editable and params.user.roles.includes('maker') %}
```
[the remark this answers is not in the document]
Comment (AlexBramhill) [reply portal/templates/contract/components/loan-link.njk]: am keeping these separate as I feel the idea of "this is what the data allows" and "this is what the role of the current user allows" should be kept separate and visible as two different checks

File: portal/templates/contract/components/loan-transactions-table.njk (hunk 28 above)
Comment (AlexBramhill) [reply portal/templates/contract/components/loan-transactions-table.njk]: updated

File: portal/templates/contract/components/loan-link.njk [hunk 32]
```diff
@@ -1,27 +1,33 @@
-{% macro render(params, loan) %}
+{% macro render(params2) %}
```
[the remark this answers is not in the document]
Comment (AlexBramhill) [reply portal/templates/contract/components/loan-link.njk]: done!

File: portal/templates/contract/components/issue-or-delete-facility-link.njk (hunk 26 above)
[the remark this answers is not in the document]
Comment (AlexBramhill) [reply portal/templates/contract/components/issue-or-delete-facility-link.njk:23]: I hope that the refactor makes this slightly more readable

File: portal/templates/contract/components/issue-or-delete-facility-link.njk (hunk 25 above)
Comment (AlexBramhill) [reply portal/templates/contract/components/issue-or-delete-facility-link.njk:29]: I think this issue is bigger than this ticket -- we have a lot of complex values being calculated inside nunjucks components -- I'd like to have a think about the best way to capture these (I'm sure there's a lot of overlap with these too).

**Comment by AlexBramhill** [comment]
I've left changing to consts (exception is portal) for later tickets as is best to do this all in one go for a service.

File: e2e-tests/portal/cypress/e2e/journeys/maker/facilities/MIA-accepted-deal-status-maker-cannot-resubmit-until-confirmed-all-issued-facilities-cover-start-date/MIA-accepted-deal-status-maker-cannot-resubmit-until-confirmed-all-issued-facilities-cover-start-date.spec.js [hunk 33]
```diff
@@ -222,7 +224,7 @@ context('Given a deal that has `Accepted` status with Issued, Unissued, Uncondit
     secondUnconditionalSubmittedLoanRow.changeOrConfirmCoverStartDateLink().click();
     pages.facilityConfirmCoverStartDate.needToChangeCoverStartDateNo().click();
 
-    coverStartDateInputsShouldNotBeVisibile();
+    coverStartDateInputsShouldNotBeVisable();
```
Comment (lmarrai-sw) [inline e2e-tests/portal/cypress/e2e/journeys/maker/facilities/MIA-accepted-deal-status-maker-cannot-resubmit-until-confirmed-all-issued-facilities-cover-start-date/MIA-accepted-deal-status-maker-cannot-resubmit-until-confirmed-all-issued-facilities-cover-start-date.spec.js:227]:
```suggestion
    coverStartDateInputsShouldNotBeVisible();
```

File: e2e-tests/portal/cypress/e2e/journeys/maker/facilities/MIA-accepted-deal-status-maker-cannot-resubmit-until-confirmed-all-issued-facilities-cover-start-date/MIA-accepted-deal-status-maker-cannot-resubmit-until-confirmed-all-issued-facilities-cover-start-date.spec.js [hunk 34]
```diff
@@ -208,7 +210,7 @@ context('Given a deal that has `Accepted` status with Issued, Unissued, Uncondit
 
     secondIssuedSubmittedBondRow.changeOrConfirmCoverStartDateLink().click();
     pages.facilityConfirmCoverStartDate.needToChangeCoverStartDateNo().click();
-    coverStartDateInputsShouldNotBeVisibile();
+    coverStartDateInputsShouldNotBeVisable();
```
Comment (lmarrai-sw) [inline e2e-tests/portal/cypress/e2e/journeys/maker/facilities/MIA-accepted-deal-status-maker-cannot-resubmit-until-confirmed-all-issued-facilities-cover-start-date/MIA-accepted-deal-status-maker-cannot-resubmit-until-confirmed-all-issued-facilities-cover-start-date.spec.js:213]:
```suggestion
    coverStartDateInputsShouldNotBeVisible();
```

File: e2e-tests/portal/cypress/e2e/journeys/maker/facilities/MIA-accepted-deal-status-maker-cannot-resubmit-until-confirmed-all-issued-facilities-cover-start-date/MIA-accepted-deal-status-maker-cannot-resubmit-until-confirmed-all-issued-facilities-cover-start-date.spec.js [hunk 35]
```diff
@@ -46,7 +48,7 @@ context('Given a deal that has `Accepted` status with Issued, Unissued, Uncondit
     });
   });
 
-  const coverStartDateInputsShouldNotBeVisibile = () => {
+  const coverStartDateInputsShouldNotBeVisable = () => {
```
Comment (lmarrai-sw) [inline e2e-tests/portal/cypress/e2e/journeys/maker/facilities/MIA-accepted-deal-status-maker-cannot-resubmit-until-confirmed-all-issued-facilities-cover-start-date/MIA-accepted-deal-status-maker-cannot-resubmit-until-confirmed-all-issued-facilities-cover-start-date.spec.js:51]:
```suggestion
  const coverStartDateInputsShouldNotBeVisible = () => {
```

File: e2e-tests/portal/cypress/e2e/journeys/read-only/deal/cannot-modify-an-unissued-deal.spec.js [hunk 36]
```diff
@@ -12,7 +12,7 @@ context('A read-only role viewing a bond that can be issued', () => {
     bonds: [],
   };
 
-  before(() => {
+  beforeEach(() => {
```
Comment (lmarrai-sw) [inline e2e-tests/portal/cypress/e2e/journeys/read-only/deal/cannot-modify-an-unissued-deal.spec.js]: Was this change deliberate (even though we delete the deals in an `after` and not an `afterEach`)?

File: e2e-tests/portal/cypress/e2e/journeys/skip-link.spec.js [hunk 37]
```diff
@@ -13,7 +13,7 @@ context('Skip link should take user to the main content of a page', () => {
     // unfortunately, cypress currently does not natively support keyboard tabbing.
     // therefore, when tabbing to display the 'skip link',
     // the skip link does not 'appear' as it would in a regular browser outside of cypress.
-    // we therefore cannot reliably test if the element is 'visibile' or 'in viewport'.
+    // we therefore cannot reliably test if the element is 'visable' or 'in viewport'.
```
Comment (lmarrai-sw) [inline e2e-tests/portal/cypress/e2e/journeys/skip-link.spec.js:16]:
```suggestion
    // we therefore cannot reliably test if the element is 'visible' or 'in viewport'.
```

File: external-api/.eslintrc.js [hunk 38]
```diff
@@ -69,5 +69,11 @@ module.exports = {
     '@typescript-eslint/no-unused-vars': 'off',
     '@typescript-eslint/ban-ts-comment': 'off',
     '@typescript-eslint/no-explicit-any': 'off',
+    "no-use-before-define": [
+      "error",
+      {
+        "functions": false
+      }
+    ]
```
Comment (lmarrai-sw) [inline external-api/.eslintrc.js]:
```suggestion
    'no-use-before-define': [
      'error',
      {
        functions: false
      },
    ],
```

**Review by lmarrai-sw — APPROVED** [review]

File: e2e-tests/portal/cypress/e2e/journeys/read-only/deal/cannot-modify-an-unissued-deal.spec.js (hunk 36 above)
Comment (AlexBramhill) [reply e2e-tests/portal/cypress/e2e/journeys/read-only/deal/cannot-modify-an-unissued-deal.spec.js]: Ah added to before each too -- so it does nothing persistent to the test db

**Review by lmarrai-sw — APPROVED** [review]

File: e2e-tests/portal/cypress/fixtures/users.js (hunk 18 above)
Comment (abhi-markan) [reply e2e-tests/portal/cypress/fixtures/users.js:4]: Sounds sensible, cannot see `role` `maker` as constant, not sure if it has been pushed?

**Review by abhi-markan — CHANGES_REQUESTED** [review]
Looks good overall, just few pending constant comments to address please.

File: portal/templates/contract/components/loan-transactions-table.njk (hunk 29 above)
Comment (AlexBramhill) [reply portal/templates/contract/components/loan-transactions-table.njk:75]: I think this issue is bigger than this ticket -- we have a lot of complex values being calculated inside nunjucks components -- I'd like to have a think about the best way to capture these (I'm sure there's a lot of overlap with these too).

**Comment by AlexBramhill** [comment]
Should be updated now thanks :)  -- As mentioned a lot of the constants etc are going to be fixed later due to complexity.

Also -- there's awider issue with complexity inside nunjucks which we should put thought into how to resolve before committing to a resolution.

File: portal/server/routes/admin/users.js [hunk 39]
```diff
@@ -7,6 +7,7 @@ const {
   generateErrorSummary,
   constructPayload,
 } = require('../../helpers');
+const { MAKER, CHECKER } = require('../../constants/roles');
```
Comment (lmarrai-sw) [inline portal/server/routes/admin/users.js:10]:
```suggestion
} = require('../../helpers');
```

**Review by lmarrai-sw — APPROVED** [review]
